### PR TITLE
Introduce USE_POLLY_ACC

### PR DESCRIPTION
**Pull request by singam-sanjay**
Introduce the USE_POLLY_ACC variable to enable Polly's GPGPU codegeneration facilities.

### CODE REVIEW

**Comment by singam-sanjay** [comment]
In https://github.com/JuliaLang/julia/pull/21736#discussion_r116689912 @tobig said:
> I understand you need the CUDA includes to compiler GPURuntime, but the LINKS flags seem to only be about linking julia with GPURuntime. For the link step, no CUDA headers should be needed.

Actually, "-lGPURuntime" isn't required in the LLVMLINK flag since I'm pulling in GPUJIT.c. I've dropped it in this pull-request.

>> +FLAGS += -I$(shell $(LLVM_CONFIG_HOST) --src-root)/tools/polly/tools

> This asks LLVM for the source directory. If LLVM is installed with a package management system, no source route will be available, right?

Yes, if Julia is to use LLVM installed via the OS's package manager the source wouldn't be available. But, Julia allows Polly to be used only with LLVM's trunk (whereas the package manager will install a stable version of LLVM). I think this is because Polly's headers aren't available through llvm-config --includedir. This is evident from [this line in src/Makefile](https://github.com/JuliaLang/julia/blob/149e5893e0ae88cb2d6bb3245b41660929d4c94a/src/Makefile#L61),
`FLAGS += -I$(shell $(LLVM_CONFIG_HOST) --src-root)/tools/polly/include`

**Comment by singam-sanjay** [comment]
@tkelman @vtjnash Can we take this forward ?

File: Make.inc [hunk 1]
```diff
@@ -239,6 +239,8 @@ BUILD_LLVM_CLANG := 0
 BUILD_LLDB := 0
 USE_POLLY := 0
 USE_POLLY_OPENMP := 0
+USE_POLLY_ACC := 0
+CUDALIB_INCLUDE_DIR := /usr/local/cuda/include
```
Comment (ararslan) [inline Make.inc]: If CUDA supports Windows, this path will be incorrect. Is there a way to detect this rather than having to hard-code it?

**Comment by staticfloat** [comment]
It would be nice to get a smidgen of documentation on what this is used for and that the option exists.  Perhaps just a comment somewhere in the makefile briefly describing what it does?

File: Make.inc (hunk 1 above)
Comment (singam-sanjay) [inline Make.inc]: @ararslan Thanks for pointing it out ! Please checkout this [diff](https://github.com/JuliaLang/julia/compare/master...singam-sanjay:USE_POLLY_ACC-withGPURuntimeHeader-CLTESTS). I'm still not sure about the correct include folder for Windows. Could you help me out with that?

Comment (ararslan) [inline Make.inc]: Sorry, but I don't really know anything about CUDA (or Polly, for that matter), so I'm not sure how best to address that. Maybe @SimonDanisch could help?

Comment (singam-sanjay) [inline Make.inc]: @ararslan NP. I see that Julia uses cygwin to build in Windows. Should paths follow the Unix notation ? with forward slashes between parent and children directories and "\" preceeding spaces ? I'm not familiar with the Cygwin environment.

**Comment by singam-sanjay** [comment]
@staticfloat I've added documentation.

**Comment by staticfloat** [comment]
While I agree with @ararslan that `CUDALIB_INCLUDE_DIR` should have a windows default setting, I don't know what that should be, and don't have the system setup to test it properly.  This is a niche enough option, and the user can override it easily enough that I say we just merge this as-is and add on a proper window default once someone actually tries to build it on Windows.

**Review by staticfloat — APPROVED** [review]

**Comment by singam-sanjay** [comment]
@staticfloat Thanks for approving the changes. I'm planning to leave CUDALIB_INCLUDE_DIR empty, for the Make.user to define it, and have the following code assign a default value incase CUDALIB_INCLUDE_DIR is still empty,
```
ifeq ($(USE_POLLY_ACC),1)
ifeq ($(CUDALIB_INCLUDE_DIR),)
ifeq ($(OS),WINNT)
ifndef CUDA_PATH
$(error Path to CUDA header files unavilable)
else
CUDALIB_INCLUDE_DIR := $(CUDA_PATH)/Include
endif
else
CUDALIB_INCLUDE_DIR := /usr/local/cuda/include
endif
$(warning Using default path to CUDA header files : ${CUDALIB_INCLUDE_DIR} )
endif
endif
```

**Comment by staticfloat** [comment]
What you have now, requiring the user to set `CUDALIB_INCLUDE_DIR` is fine, IMO.  I would change the error message to say `CUDALIB_INCLUDE_DIR not set: If compiling with USE_POLLY_ACC, user must provide the path to CUDA header files`

**Comment by staticfloat** [comment]
Thanks for this contribution @singam-sanjay!  I'm glad you're pushing forward on this direction and welcome to the Julia project!

**Comment by singam-sanjay** [comment]
Thanks @staticfloat @ararslan for your time and help !!

**Pull request merged**